File: detector/background.py
"""
background.py  --  Per-cell background model for the VL53L5CX doorway node.

PORTABILITY CONTRACT
--------------------
This module is the GOLDEN REFERENCE that gets ported to C/ESP-IDF in a later
phase. Keep it that way:
  * plain Python only -- NO numpy, NO pandas, NO list comprehensions that
    hide loops you'd have to unroll in C.
  * fixed-size arrays of length N_CELLS. No dynamic growth in the hot path.
  * integer/float math that maps 1:1 to C.
The harness (replay_background.py) is allowed to use numpy/pandas. This file
is not. If you reach for numpy here, stop -- it belongs in the harness.

DATA MODEL
----------
A frame = two arrays of length 64:
    dist[i]   -- distance in mm for cell i        (int)
    status[i] -- ST range status code for cell i  (int)
Trusted status codes are {5, 9, 10, 255} (Phase 1 finding #1).
Cell HOT_PIXEL is masked unconditionally (Phase 1 finding #4).
"""

GRID = 8
N_CELLS = GRID * GRID          # 64

VALID_STATUS = (5, 9, 10, 255)  # Phase 1 finding #1
HOT_PIXELS = (5,)               # flat index of cell (row=0,col=5) -> 0*8+5 = 5
                                # kept as DATA, not an `if` -- swap to per-node
                                # auto-detection later by changing this tuple.

# ---- tuning knobs (you will sweep these against the fixtures) -------------
BOOTSTRAP_FRAMES = 50    # frames assumed empty at startup to seed bg
ALPHA = 0.01             # EMA learning rate; 1/ALPHA ~= frames to adapt
                         #   0.01 -> ~100 frames -> ~10s at 10Hz
K_OCCUPY = 6.0           # occupied if dev > K_OCCUPY * mad   (declare)
K_CLEAR  = 4.0           # clears  if dev < K_CLEAR  * mad   (hysteresis)
MAD_FLOOR_MM = 15.0      # min noise floor so quiet cells aren't hair-trigger
MIN_OCCUPIED_CELLS = 2   # frame is "occupied" if >= this many cells fire
# ---------------------------------------------------------------------------


def rc_to_idx(row, col):
    """Map (row, col) -> flat index. Keep one convention everywhere."""
    return row * GRID + col


def _trusted(status, idx):
    """Is cell idx usable this frame?"""
    if idx in HOT_PIXELS:
        return False
    return status[idx] in VALID_STATUS


def _median(values):
    """Median of a list. Plain, no numpy. values must be non-empty."""
    s = sorted(values)
    n = len(s)
    mid = n // 2
    if n % 2 == 1:
        return float(s[mid])
    return 0.5 * (s[mid - 1] + s[mid])


class BackgroundModel:
    """
    Holds per-cell background estimate and noise estimate.
    Lifecycle:
        m = BackgroundModel()
        for first BOOTSTRAP_FRAMES frames:   m.bootstrap_add(dist, status)
        m.bootstrap_finalize()
        then per frame:                      occ, dev = m.process(dist, status)
    """

    def __init__(self):
        self.bg = [0.0] * N_CELLS          # background estimate (mm)
        self.mad = [0.0] * N_CELLS         # per-cell noise estimate (mm)
        self.calibrated = [False] * N_CELLS
        self._boot = [[] for _ in range(N_CELLS)]  # bootstrap sample buffers
        self._ready = False
        # per-cell hysteresis latch: is this cell currently "occupied"?
        self._cell_hot = [False] * N_CELLS

    # ---- bootstrap ---------------------------------------------------------
    def bootstrap_add(self, dist, status):
        """Collect one frame's trusted samples into per-cell buffers."""
        for i in range(N_CELLS):
            if _trusted(status, i):
                self._boot[i].append(dist[i])

    def bootstrap_finalize(self):
        """
        Turn collected samples into bg (median) + mad (median abs deviation).
        Median, not mean: tolerates a person wandering through during calib.
        A cell with no trusted samples stays uncalibrated (handled as no-data).
        """
        for i in range(N_CELLS):
            samples = self._boot[i]
            if len(samples) == 0:
                self.calibrated[i] = False
                continue
            med = _median(samples)
            abs_dev = [abs(v - med) for v in samples]
            mad = _median(abs_dev)
            self.bg[i] = med
            self.mad[i] = max(mad, MAD_FLOOR_MM)
            self.calibrated[i] = True
        self._boot = [[] for _ in range(N_CELLS)]  # free memory
        self._ready = True

    # ---- per-frame ---------------------------------------------------------
    def process(self, dist, status):
        """
        Returns (occupied_cells, dev) where:
          occupied_cells -- list of flat indices currently occupied
          dev            -- length-64 list, deviation in mm (0.0 where no-data)
        Side effect: gated EMA update of bg on frames deemed empty.
        """
        if not self._ready:
            raise RuntimeError("call bootstrap_finalize() first")

        dev = [0.0] * N_CELLS
        occupied = []

        for i in range(N_CELLS):
            if not self.calibrated[i] or not _trusted(status, i):
                self._cell_hot[i] = False
                continue
            d = self.bg[i] - dist[i]      # positive = closer than bg = object
            dev[i] = d
            hi = K_OCCUPY * self.mad[i]
            lo = K_CLEAR * self.mad[i]
            # hysteresis: high bar to turn on, low bar to turn off
            if self._cell_hot[i]:
                if d < lo:
                    self._cell_hot[i] = False
            else:
                if d > hi:
                    self._cell_hot[i] = True
            if self._cell_hot[i]:
                occupied.append(i)

        frame_occupied = len(occupied) >= MIN_OCCUPIED_CELLS

        # GATE: only learn background from frames that look empty.
        # This is the fix for absorbing a stationary person (stop-in-doorway).
        if not frame_occupied:
            for i in range(N_CELLS):
                if self.calibrated[i] and _trusted(status, i):
                    self.bg[i] = (1.0 - ALPHA) * self.bg[i] + ALPHA * dist[i]

        return occupied, dev

File: doorway-node-01/src/detection/background.c
/*
 * background.c -- faithful C port of detector/background.py.
 * See background.h for the contract. Pure C, no ESP-IDF / FreeRTOS / I2C.
 */
#include "background.h"

#include <stdlib.h>   /* qsort */
#include <math.h>     /* fabsf */

/* ---- masks / trusted status, kept as DATA (Phase 1 findings #1, #4) ----- */
static const int VALID_STATUS[] = {5, 9, 10, 255};
static const int N_VALID_STATUS = (int)(sizeof(VALID_STATUS) / sizeof(VALID_STATUS[0]));

static const int HOT_PIXELS[] = {1, 5};  /* (0,1) bimodal flicker, (0,5) stuck hot */
static const int N_HOT_PIXELS = (int)(sizeof(HOT_PIXELS) / sizeof(HOT_PIXELS[0]));

bool bg_trusted(const int status[BG_N_CELLS], int idx)
{
    for (int h = 0; h < N_HOT_PIXELS; h++) {
        if (idx == HOT_PIXELS[h]) {
            return false;
        }
    }
    int s = status[idx];
    for (int v = 0; v < N_VALID_STATUS; v++) {
        if (s == VALID_STATUS[v]) {
            return true;
        }
    }
    return false;
}

/* ---- median helpers (plain, no numpy) ---------------------------------- */
static int cmp_int(const void *a, const void *b)
{
    int x = *(const int *)a, y = *(const int *)b;
    return (x > y) - (x < y);
}

static int cmp_float(const void *a, const void *b)
{
    float x = *(const float *)a, y = *(const float *)b;
    return (x > y) - (x < y);
}

/* median of an int array (sorts it in place); n must be >= 1 */
static float median_int(int *vals, int n)
{
    qsort(vals, (size_t)n, sizeof(int), cmp_int);
    int mid = n / 2;
    if (n % 2 == 1) {
        return (float)vals[mid];
    }
    return 0.5f * ((float)vals[mid - 1] + (float)vals[mid]);
}

/* median of a float array (sorts it in place); n must be >= 1 */
static float median_float(float *vals, int n)
{
    qsort(vals, (size_t)n, sizeof(float), cmp_float);
    int mid = n / 2;
    if (n % 2 == 1) {
        return vals[mid];
    }
    return 0.5f * (vals[mid - 1] + vals[mid]);
}

/* ---- connectivity ------------------------------------------------------ */
int det_largest_blob(const bool occ[BG_N_CELLS], int out_cells[BG_N_CELLS])
{
    bool seen[BG_N_CELLS] = {false};
    int  stack[BG_N_CELLS];
    int  comp[BG_N_CELLS];
    int  best_n = 0;

    for (int start = 0; start < BG_N_CELLS; start++) {
        if (!occ[start] || seen[start]) {
            continue;
        }
        int comp_n = 0;
        int top = 0;
        stack[top++] = start;
        seen[start] = true;
        while (top > 0) {
            int cell = stack[--top];
            comp[comp_n++] = cell;
            int row = cell / BG_GRID;
            int col = cell % BG_GRID;
            /* 8-connected neighborhood (incl. diagonals) -- finding #5 */
            for (int dr = -1; dr <= 1; dr++) {
                for (int dc = -1; dc <= 1; dc++) {
                    if (dr == 0 && dc == 0) {
                        continue;
                    }
                    int nr = row + dr;
                    int nc = col + dc;
                    if (nr < 0 || nr >= BG_GRID || nc < 0 || nc >= BG_GRID) {
                        continue;
                    }
                    int nb = nr * BG_GRID + nc;
                    if (occ[nb] && !seen[nb]) {
                        seen[nb] = true;
                        stack[top++] = nb;
                    }
                }
            }
        }
        if (comp_n > best_n) {
            best_n = comp_n;
            for (int i = 0; i < comp_n; i++) {
                out_cells[i] = comp[i];
            }
        }
    }
    return best_n;
}

/* ---- lifecycle --------------------------------------------------------- */
void bg_init(background_model_t *m)
{
    for (int i = 0; i < BG_N_CELLS; i++) {
        m->bg[i] = 0.0f;
        m->mad[i] = 0.0f;
        m->calibrated[i] = false;
        m->cell_hot[i] = false;
        m->boot_count[i] = 0;
    }
    m->ready = false;
}

void bg_bootstrap_add(background_model_t *m,
                      const int dist[BG_N_CELLS],
                      const int status[BG_N_CELLS])
{
    for (int i = 0; i < BG_N_CELLS; i++) {
        if (bg_trusted(status, i)) {
            int c = m->boot_count[i];
            if (c < BG_BOOTSTRAP_FRAMES) {
                m->boot[i][c] = dist[i];
                m->boot_count[i] = c + 1;
            }
        }
    }
}

void bg_bootstrap_finalize(background_model_t *m)
{
    for (int i = 0; i < BG_N_CELLS; i++) {
        int n = m->boot_count[i];
        if (n == 0) {
            m->calibrated[i] = false;
            continue;
        }
        float med = median_int(m->boot[i], n);  /* sorts boot[i] in place */
        float devs[BG_BOOTSTRAP_FRAMES];
        for (int j = 0; j < n; j++) {
            devs[j] = fabsf((float)m->boot[i][j] - med);
        }
        float mad = median_float(devs, n);
        m->bg[i] = med;
        m->mad[i] = (mad > BG_MAD_FLOOR_MM) ? mad : BG_MAD_FLOOR_MM;
        m->calibrated[i] = true;
    }
    /* free the bootstrap buffers (reset counts; memory is static here) */
    for (int i = 0; i < BG_N_CELLS; i++) {
        m->boot_count[i] = 0;
    }
    m->ready = true;
}

void bg_process(background_model_t *m,
                const int dist[BG_N_CELLS],
                const int status[BG_N_CELLS],
                bool occupied[BG_N_CELLS],
                float dev[BG_N_CELLS])
{
    /* Caller must have finalized. (No exceptions in C: just no-op-safe guard.) */
    for (int i = 0; i < BG_N_CELLS; i++) {
        dev[i] = 0.0f;
        occupied[i] = false;
    }
    if (!m->ready) {
        return;
    }

    for (int i = 0; i < BG_N_CELLS; i++) {
        if (!m->calibrated[i] || !bg_trusted(status, i)) {
            m->cell_hot[i] = false;
            continue;
        }
        float d = m->bg[i] - (float)dist[i];  /* positive = closer than bg = object */
        dev[i] = d;
        float hi = BG_K_OCCUPY * m->mad[i];
        float lo = BG_K_CLEAR * m->mad[i];
        /* hysteresis: high bar to turn on, low bar to turn off */
        if (m->cell_hot[i]) {
            if (d < lo) {
                m->cell_hot[i] = false;
            }
        } else {
            if (d > hi) {
                m->cell_hot[i] = true;
            }
        }
        if (m->cell_hot[i]) {
            occupied[i] = true;
        }
    }

    /* BLOB GATE: a frame is occupied only if its largest connected blob is big
     * enough. A lone size-1 cell (e.g. the bimodal (0,1) cell) never trips it,
     * so bg keeps learning everywhere else. */
    int scratch[BG_N_CELLS];
    int largest = det_largest_blob(occupied, scratch);
    bool frame_occupied = largest >= BG_MIN_BLOB_CELLS;

    /* GATE: learn bg only from frames that look empty, AND never fold a cell
     * into bg while that cell is currently flagged hot. */
    if (!frame_occupied) {
        for (int i = 0; i < BG_N_CELLS; i++) {
            if (m->calibrated[i] && bg_trusted(status, i) && !m->cell_hot[i]) {
                m->bg[i] = (1.0f - BG_ALPHA) * m->bg[i] + BG_ALPHA * (float)dist[i];
            }
        }
    }
}
